lib/autoproj/daemon/buildbot.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require 'autoproj'
+require 'net/http'
+require 'uri'
+require 'json'
+
+module Autoproj
+    module Daemon
+        # Buildbot integration class
+        class Buildbot
+            # @return [Autoproj::Workspace]
+            attr_reader :ws
+
+            HEADER = { 'Content-Type' => 'application/json' }.freeze
+            BODY = {
+                method: 'force',
+                jsonrpc: '2.0',
+                id: 1,
+                params: {
+                }
+            }.freeze
+
+            # @param [Autoproj::Workspace] workspace
+            def initialize(workspace)
+                @ws = workspace
+            end
+
+            # @param [Hash] options
+            # @return [Hash]
+            def validate_options(options = {})
+                options[:branch] ||= 'master'
+                options
+            end
+
+            # @param [Hash] options
+            # @return [Hash]
+            def body(**options)
+                body = BODY
+                body[:params].merge!(validate_options(options))
```

Comment:
You are modifying BODY in place.

~~~
body = BODY.merge(
    params: BODY[:params].merge(branch: branch).merge(**options)
)
~~~

Reply:
Good catch... I wonder why this didn't raise anything even though `BODY` is frozen

Reply:
`freeze` is not recursive, so modifying BODY would have raised, but BODY[:params] is another object.